### ISSUE TITLE
Ask a question about decrypt

### ISSUE THREAD

**Issue by WcLyic** [morkt/GARbro#70]
Can you share the method to find out the decrypt function of the KrkrZ?
I want to unpack and repack something with some old games, and want to create the CryptClass by myself.

**Comment by morkt** [comment]
first i should note that with current GARbro formats infrastructure custom encryption methods couldn't be easily made visible by GUI component unless i include them into repository myself.

that said, to find out decryption method you should be familiar with assembly language and reverse engineering techniques in general. recommended tools: hiew and IDA+Hex-Rays.

## Simple extraction filter ##

locate TPM plugin that contains filter, usually it's a file named either `{GAME TITLE}.tpm` or `xp3dec.tpm`. it's a DLL that contains two entry points: `V2Link` and `V2Unlink`. follow `V2Link` subroutine and find an indirect call to `TVPSetXP3ArchiveExtractionFilter`. first argument to that function will be extraction filter subroutine address.

filter is called when file is being read from archive with single argument that is a pointer to the following structure:

    struct tTVPXP3ExtractionFilterInfo
    {
        uint32  SizeOfSelf; // structure size of tTVPXP3ExtractionFilterInfo itself (== 0x18)
        uint64  Offset;     // offset of the buffer data in uncompressed stream position
        void*   Buffer;     // target data buffer
        uint32  BufferSize; // buffer size in bytes pointed by "Buffer"
        uint32  FileHash;   // hash value of the file
    };

that subroutine should be implemented as `Decrypt` method of GARbro XP3 `ICrypt` class implementation. if you have IDA Hex-Rays decompiler, translating it from assembler is a pretty straight forward process.

## Encrypted filters ##

recent games are shipped with encrypted extraction filters that decrypt themselves during initialization.  simplest way to circumvent such "protection" is to attach to a running game with debugger and find extraction filter in process memory after it has been decrypted.

## `CXDEC` filters ##

these filters compose decryption functions 'on the fly' by randomly combining predefined chunks of machine code. implemented by `CxEncryption` class in GARbro. it's a bit complicated, and i don't have much time to explain it right now, so maybe later.

**Comment by WcLyic** [comment]
Thank you for explain three types of encryption. And I think what i want to crack is the second types.
I have tried to find the key function but it is too hard for me without symbols and i also have tried to debug the demo program of krkrz but it also lack of pdb. Are you have some eigen codes to exactly follow in ?

**Comment by morkt** [comment]
what's "eigen codes"? and yes, debugging code without symbols is not an easy process. like i said, you should have a general understanding of assembler and what certain C/C++ language constructs will look like in machine codes. although it's a bit easier for Kirikiri since its source code is open. for example, take a look at [XP3Archive.cpp](https://github.com/krkrz/krkrz/blob/master/base/XP3Archive.cpp). extraction filter is called from `tTVPXP3ArchiveStream::Read` method, so you should locate its code. usually C++ compiler/linker translates methods into executable file in the same order they appear in source, so locating any method from that file will lead you to `Read`. particularly, `TVPGetXP3ArchiveOffset` could easily be spotted since it has hard-coded sequence `XP3\x0d\x0a` referenced. another anchor is `tTVPSegmentData::SetData` since it calls zlib `uncompress` function, and zlib methods have a distinguishable signature no matter what application they are used in.

once `tTVPXP3ArchiveStream::Read` is located, you could get an address of extraction filter subroutine. some games use a modified Kirikiri version that applies extraction filter elsewhere, but it's still possible to follow it by placing breakpoints and tracing a call stack.

**Comment by WcLyic** [comment]
Thank you very much and i have successed in decrypting the first game (Purely x Cation).
```csharp
public override void Decrypt(Xp3Entry entry, long offset, byte[] buffer, int pos, int count)
{
    table = Convert.FromBase64String("FUw1sZQAAAAGsGy5+QAAAC92WgDQAAAAQHQv7r8AAAAduocD4gAAAP2e09kCAAAA6Fhu/xcAAAD7Sim5BAAAADhG19DHAAAAsMeqoE8AAABbhqDOpAAAABTbYRzrAAAAxw0V7zgAAADRk6ZDLgAAAIApDXF/AAAAOiq6x8UAAAAO2q398QAAAKwqj0lTAAAAKFuF/9cAAAD6lJWNBQAAAHrOElqFAAAA31ylbyAAAADIX4oqNwAAAIHSx+Z+AAAAby+ERZAAAAC4Uzh2RwAAAKZvhgNZAAAAfjCmVYEAAABSdwnhrQAAAMOA+I08AAAAApB7Yv0AAAANvbBp8gAAAEtmXXG0AAAAS5vXsbQAAACmb93JWQAAAN2sqVQiAAAAzWoezjIAAACUxTbOawAAAHLBJaqNAAAAovI3gF0AAACPyRE0cAAAAAkliev2AAAA/QAzMwIAAAAHCFfK+AAAAFN+SM+sAAAAE6DmROwAAAA/C3F+wAAAAPX8WmgKAAAA7bu3xRIAAACwePBZTwAAAH8UXAaAAAAAUAE6W68AAAD8sGzbAwAAAKzfVf1TAAAAMmOSos0AAAC3bsa0SAAAAMlEcBo2AAAAySgfijYAAAA4oYJcxwAAAGlbHGiWAAAAQc2fu74AAAA=");
    for (int i = 0; i < 0x7A; i++)
    {
        if (entry.Hash == BitConverter.ToUInt32(table, i))
        {
            for (int j = 0; j < count; j++)
                buffer[j] ^= table[i + 4];
            return;
        }
    }
    uint hash = entry.Hash;
    byte[] key1 = new byte[32];
    for (int k = 0; k < 8; k++)
    {
        uint t = 0;
        if ((hash & 1) != 0)
            t = hash | 0x80000000;
        else
            t = hash & 0x7FFFFFFF;
        Array.Copy(BitConverter.GetBytes(t), 0, key1, k * 4, 4);
        hash = t >> 1;
    }
    byte[] key2 = new byte[1024];
    for (int l = 0; l < 32; l++)
    {
        var t = key1[l] ^ 0xFF;
        for (int m = 0; m < 32; m++)
        {
            key2[l * 32 + m] = (byte)(t ^ key1[m]);
        }
    }
    for (int i = 0; i < count; i++)
        buffer[pos + i] ^= key2[(offset + i) & 0x3FF];
}
```
And I tested the decrypt by modifying HbikiCrypt and used the "SerializeScheme" to add the option of my Crypt. XD

**Comment by UlyssesWu** [comment]
(Maybe Unrelated)
@WcLyic I have heard that PxC uses Emote. Now that you have succeeded, would you like to share PSBs with me? I just would like to have more samples for development of `FreeMote`, you know. Thanks.

**Comment by WcLyic** [comment]
@UlyssesWu  You can just update the GARbro and export by yourself because @morkt  had added it in.

**Comment by UlyssesWu** [comment]
@WcLyic Thanks for the info. But in fact I'm not interested in that game and don't want to download it actually...

**Comment by WcLyic** [comment]
@UlyssesWu I sure when you finish donwload and I haven`t uploading completly. (my upload speed:3Mbps)  XD

**Comment by UlyssesWu** [comment]
😭 Although PSB files may be 70MB for each, they can be quite small (usually 7MB) after compression because there are much redundance in them. 
BTW, I'm also curious in XP3 unpacking topic since I hope I can extract and repack [Emote Editor](http://emote.mtwo.co.jp/academic-guide.php) someday... And I have made a little improvement for [KirikiriSharp](https://github.com/Project-AZUSA/KirikiriSharp) (still super WIP).

**Comment by WcLyic** [comment]
For both Pretty x Cation and Pretty x Cation 2 :smile:
```csharp
public override void Decrypt (Xp3Entry entry, long offset, byte[] buffer, int pos, int count)
{
    byte key = (byte)((entry.Hash >> 12) & 0xFF);
    for (int i = 0; i < count; i++)
    {
        if (offset + i > 4)
            buffer[pos + i] ^= key;
    }
}
```

**Comment by morkt** [comment]
looks like it's already implemented as `AppliqueCrypt`.

**Comment by WcLyic** [comment]
Yes, the groups belong to Akabei Soft2 use the same or similar cryption in the close development time. :laughing:
